From 50466d97d6cd589a2ef6d59e719ae2cebcf98ee1 Mon Sep 17 00:00:00 2001
From: Ziad Elhanafy <ziad.elhanafy@arm.com>
Date: Thu, 8 Aug 2024 16:25:07 +0100
Subject: [PATCH] Platform: Kronos: Configure RSS MPU

The RSS MPU needs to be configured with the Secure RSS
ATU data region as device memory in order to access
ATU-mapped devices.

Upstream-Status: Pending
Signed-off-by: Ziad Elhanafy <ziad.elhanafy@arm.com>
---
 .../ext/target/arm/rss/kronos/CMakeLists.txt  |  2 +
 .../target/arm/rss/kronos/bl2/boot_hal_bl2.c  | 48 ++++++++++++++++++-
 2 files changed, 49 insertions(+), 1 deletion(-)

diff --git a/platform/ext/target/arm/rss/kronos/CMakeLists.txt b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
index 1f2d45247..a98da7034 100644
--- a/platform/ext/target/arm/rss/kronos/CMakeLists.txt
+++ b/platform/ext/target/arm/rss/kronos/CMakeLists.txt
@@ -134,6 +134,7 @@ target_sources(platform_bl2
         fw_update_agent/uefi_capsule_parser.c
         fw_update_agent/uefi_fmp.c
         ${PLATFORM_DIR}/ext/common/tfm_hal_spm_logdev_peripheral.c
+        ${PLATFORM_DIR}/ext/target/arm/drivers/mpu/armv8m/mpu_armv8m_drv.c
 )
 
 target_compile_definitions(platform_bl2
@@ -159,6 +160,7 @@ target_include_directories(platform_bl2
         ${PLATFORM_DIR}/ext/target/arm/drivers/usart/pl011
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/common
         ${PLATFORM_DIR}/ext/target/arm/drivers/flash/strata
+        ${PLATFORM_DIR}/ext/target/arm/drivers/mpu/armv8m/
         ${PLATFORM_DIR}/ext/target/arm/rss/common/fip_parser
         io
         soft_crc
diff --git a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
index e773be0c6..3730e71d5 100644
--- a/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
+++ b/platform/ext/target/arm/rss/kronos/bl2/boot_hal_bl2.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, Arm Limited. All rights reserved.
+ * Copyright (c) 2023-2024, Arm Limited. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  *
@@ -19,6 +19,7 @@
 #include "host_base_address.h"
 #include "mhu.h"
 #include "mhu_v3_x.h"
+#include "mpu_armv8m_drv.h"
 #include "mscp_mailbox.h"
 #include "platform_base_address.h"
 #include "platform_regs.h"
@@ -327,6 +328,46 @@ int32_t tower_nci_periph_init(void) {
     return 0;
 }
 
+static int32_t configure_mpu(void)
+{
+    const struct mpu_armv8m_dev_t dev_mpu_s = { MPU_BASE };
+    const struct mpu_armv8m_region_cfg_t region_cfg[] = {
+        {
+            0,
+            HOST_ACCESS_BASE_S,
+            HOST_ACCESS_LIMIT_S,
+            MPU_ARMV8M_MAIR_ATTR_DEVICE_IDX,
+            MPU_ARMV8M_XN_EXEC_NEVER,
+            MPU_ARMV8M_AP_RW_PRIV_UNPRIV,
+            MPU_ARMV8M_SH_NONE
+        },
+        {
+            1,
+            HOST_ACCESS_BASE_NS,
+            HOST_ACCESS_LIMIT_NS,
+            MPU_ARMV8M_MAIR_ATTR_DEVICE_IDX,
+            MPU_ARMV8M_XN_EXEC_NEVER,
+            MPU_ARMV8M_AP_RW_PRIV_UNPRIV,
+            MPU_ARMV8M_SH_NONE
+        },
+    };
+
+    if (mpu_armv8m_clean(&dev_mpu_s) != MPU_ARMV8M_OK) {
+        return ARM_DRIVER_ERROR;
+    }
+
+    if ((mpu_armv8m_region_enable(&dev_mpu_s, &region_cfg[0]) != MPU_ARMV8M_OK) ||
+        (mpu_armv8m_region_enable(&dev_mpu_s, &region_cfg[1]) != MPU_ARMV8M_OK)) {
+        return ARM_DRIVER_ERROR;
+    }
+
+    if (mpu_armv8m_enable(&dev_mpu_s, PRIVILEGED_DEFAULT_ENABLE, HARDFAULT_NMI_ENABLE) != MPU_ARMV8M_OK) {
+        return ARM_DRIVER_ERROR;
+    }
+
+    return ARM_DRIVER_OK;
+}
+
 int32_t boot_platform_init(void)
 {
     int32_t result;
@@ -334,6 +375,11 @@ int32_t boot_platform_init(void)
 
     read_chip_id();
 
+    result = configure_mpu();
+    if (result != ARM_DRIVER_OK) {
+        return 1;
+    }
+
     result = tower_nci_sysctrl_init();
     if (result != ARM_DRIVER_OK) {
         return 1;
-- 
2.34.1

